packages/devtools_app/lib/src/flutter/banner_messages.dart

```diff
@@ -20,36 +21,61 @@ const _profileGranularityDocsUrl =
     'https://flutter.dev/docs/development/tools/devtools/performance#profile-granularity';
 
 class BannerMessagesController {
-  final _messages = <DevToolsScreenType, List<BannerMessage>>{};
+  final _messages = <DevToolsScreenType, ValueNotifier<List<BannerMessage>>>{};
   final _dismissedMessageKeys = <Key>{};
 
   void addMessage(BannerMessage message) {
-    assert(!isMessageVisible(message));
-    final messages = messagesForScreen(message.screenType);
-    messages.add(message);
+    if (isMessageDismissed(message) || isMessageVisible(message)) return;
+    final messages = _messagesForScreen(message.screenType);
+    messages.value.add(message);
+    // ignore: invalid_use_of_visible_for_testing_member, invalid_use_of_protected_member
+    messages.notifyListeners();
   }
 
   void removeMessage(BannerMessage message, {bool dismiss = false}) {
-    final messages = messagesForScreen(message.screenType);
-    messages.remove(message);
     if (dismiss) {
       assert(!_dismissedMessageKeys.contains(message.key));
       _dismissedMessageKeys.add(message.key);
     }
+    final messages = _messagesForScreen(message.screenType);
+    messages.value.remove(message);
+    // ignore: invalid_use_of_visible_for_testing_member, invalid_use_of_protected_member
+    messages.notifyListeners();
+  }
+
+  void removeMessageByKey(Key key, DevToolsScreenType screenType) {
+    final currentMessages = _messagesForScreen(screenType);
+    final messageWithKey = currentMessages.value.firstWhere(
+      (m) => m.key == key,
+      orElse: () => null,
+    );
+    if (messageWithKey != null) {
+      removeMessage(messageWithKey);
+    }
   }
 
   bool isMessageDismissed(BannerMessage message) {
```

Comment:
nit: you can probably make this message @protected or private now.

Reply:
made visible for testing - here and below